source/UF/Truncations.lagda

```diff
@@ -0,0 +1,435 @@
+                → (P : ∥ X ∥[ n ] → ∥ Y ∥[ n ] → 𝓦 ̇) 
+                → (m : (u : ∥ X ∥[ n ]) → (v : ∥ Y ∥[ n ])
+                 → (P u v) is-of-hlevel n)
+                → (g : (x : X) → (y : Y) → P (∣ x ∣[ n ]) (∣ y ∣[ n ]))
+                → (x : X) → (y : Y)
+                → ∥∥ₙ-ind₂ P m g ∣ x ∣[ n ] ∣ y ∣[ n ] ＝ g x y
+  ∥∥ₙ-ind-comp₂ {𝓤} {𝓥} {𝓦} {X} {Y} {n} P m g x y =
+   ∥∥ₙ-ind₂ P m g ∣ x ∣[ n ] ∣ y ∣[ n ]     ＝⟨ I ⟩
+   ∥∥ₙ-ind (m ∣ x ∣[ n ]) (g x) ∣ y ∣[ n ]  ＝⟨ II ⟩
+   g x y                                    ∎
+    where
+     I : ∥∥ₙ-ind₂ P m g ∣ x ∣[ n ] ∣ y ∣[ n ]
+       ＝ ∥∥ₙ-ind (m ∣ x ∣[ n ]) (g x) ∣ y ∣[ n ]
+     I = happly
+          (∥∥ₙ-ind-comp (λ u → hlevel-closed-under-Π n (P u) (m u))
+           (λ x' → ∥∥ₙ-ind (m ∣ x' ∣[ n ]) (g x')) x)
+          ∣ y ∣[ n ]
+     II : ∥∥ₙ-ind (m ∣ x ∣[ n ]) (g x) ∣ y ∣[ n ] ＝ g x y
+     II = ∥∥ₙ-ind-comp (m ∣ x ∣[ n ]) (g x) y
+
+\end{code}
+
+We characterize the first couple levels of truncation (TODO: three-hlevel is a
+groupoid). 
+
+\begin{code}
+
+ zero-trunc-is-contr : {X : 𝓤 ̇ } → is-contr (∥ X ∥[ 0 ])
+ zero-trunc-is-contr = ∥∥ₙ-hlevel 0
+
+ one-trunc-is-prop : {X : 𝓤 ̇ } → is-prop (∥ X ∥[ 1 ])
+ one-trunc-is-prop = is-prop'-implies-is-prop (∥∥ₙ-hlevel 1)
+
+ two-trunc-is-set : {X : 𝓤 ̇ } → is-set (∥ X ∥[ 2 ])
+ two-trunc-is-set {𝓤} {X} {x} {y} =
+  is-prop'-implies-is-prop (∥∥ₙ-hlevel (2) x y)
+
+\end{code}
+
+We demonstrate the equivalence of one-truncation and propositional truncation:
+ ∥ X ∥₁ ≃ ∥ X ∥
+
+\begin{code}
+
+ open propositional-truncations-exist pt
+
+ one-trunc-to-prop-trunc : {X : 𝓤 ̇} → ∥ X ∥[ 1 ] → ∥ X ∥
+ one-trunc-to-prop-trunc = ∥∥ₙ-rec (is-prop-implies-is-prop' ∥∥-is-prop) ∣_∣
+
+ prop-trunc-to-one-trunc : {X : 𝓤 ̇} → ∥ X ∥ → ∥ X ∥[ 1 ]
+ prop-trunc-to-one-trunc = ∥∥-rec one-trunc-is-prop (∣_∣[ 1 ])
+
+ one-trunc-≃-prop-trunc : {X : 𝓤 ̇}
+                        → (∥ X ∥[ 1 ]) ≃ ∥ X ∥
+ one-trunc-≃-prop-trunc {𝓤} {X} =
+  logically-equivalent-props-are-equivalent one-trunc-is-prop ∥∥-is-prop
+                                            one-trunc-to-prop-trunc
+                                            prop-trunc-to-one-trunc
+
+\end{code}
+
+We provide the canonical predecessor map and show truncations are closed under
+equivalence and successive applications of truncation (TODO: other closure
+conditions (?)).
+
+\begin{code}
+ canonical-pred-map : {X : 𝓤 ̇} {n : ℕ}
+                    → ∥ X ∥[ succ n ] → ∥ X ∥[ n ]
+ canonical-pred-map {𝓤} {X} {n} x =
+  ∥∥ₙ-rec (hlevels-are-upper-closed n (∥ X ∥[ n ]) (∥∥ₙ-hlevel n))
+          (λ x → ∣ x ∣[ n ]) x
+
+ canonical-pred-map-comp : {X : 𝓤 ̇} {n : ℕ} (x : X)
+                         → canonical-pred-map (∣ x ∣[ succ n ]) ＝ (∣ x ∣[ n ])
+ canonical-pred-map-comp {𝓤} {X} {n} x =
+  ∥∥ₙ-rec-comp (hlevels-are-upper-closed n (∥ X ∥[ n ]) (∥∥ₙ-hlevel n))
+               (λ _ → ∣ _ ∣[ n ]) x
+
+ to-∼-of-maps-with-truncated-domain : {n : ℕ} {X : 𝓤 ̇} {Y : 𝓥 ̇}
```

Comment:
In this and the following two lemmas, for consistency, you probably want to move the argument `n` to appear after `X` and `Y`, at least that appears to be what you do in the other places.